### PR TITLE
Add PiecewiseChebyshev arithmetic circuit

### PR DESCRIPTION
**Pull request by anedumla**
### Summary
Adds the new feature to the arithmetic module.

Blocked on #5332

### Details and comments
Given a function as an input, find a piecewise polynomial approximation with `numpy`'s `Chebyshev` module and implement it as Pauli rotations using the `PiecewisePolynomialRotations` class.

### CODE REVIEW

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+        self._poly_r = PiecewisePolynomialPauliRotations(self.num_state_qubits,
+                                                         self.breakpoints, self.polynomials)
+
+        qr_state = self.qubits[:self.num_state_qubits]
+        qr_target = [self.qubits[self.num_state_qubits]]
+        qr_ancillas = self.qubits[self.num_state_qubits + 1:]
+
+        # Apply polynomial approximation
+        self.append(self._poly_r.to_instruction(), qr_state[:] + qr_target + qr_ancillas[:])
```
Comment (Cryoris) [inline qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:327]: Doesn't need to be a member anymore
```suggestion
        poly_r = PiecewisePolynomialPauliRotations(self.num_state_qubits,
                                                         self.breakpoints, self.polynomials)

        qr_state = self.qubits[:self.num_state_qubits]
        qr_target = [self.qubits[self.num_state_qubits]]
        qr_ancillas = self.qubits[self.num_state_qubits + 1:]

        # Apply polynomial approximation
        self.append(poly_r.to_instruction(), qr_state[:] + qr_target + qr_ancillas[:])
```

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        # define internal parameters
+        self._num_state_qubits = None
+
+        # Store parameters
+        self._f_x = f_x
+        self._degree = degree if degree is not None else 1
+        self._breakpoints = breakpoints if breakpoints is not None else [0]
+
+        self._polynomials = None
+        self._poly_r = None
```
Comment (Cryoris) [inline qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:84]: Not needed anymore I think
```suggestion
```

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            self._reset_registers(self.num_state_qubits)
+
+    @property
+    def breakpoints(self) -> List[int]:
+        """The breakpoints for the piecewise approximation.
+
+                Returns:
+                    The breakpoints for the piecewise approximation.
+        """
+        breakpoints = self._breakpoints
```
Comment (Cryoris) [inline qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:176]: We should add a check if `num_state_qubits` is None

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                Returns:
+                    The polynomials for the piecewise approximation.
+        """
+        if self.num_state_qubits is None:
+            return [[]]
+
+        n_l = 2 ** self.num_state_qubits
+
+        # Add n_l as a
+        num_intervals = len(self._breakpoints)
```
Comment (Cryoris) [inline qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:219]: Should this use `self.breakpoints` instead?

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+            # Set breakpoints if they haven't been set
+            if num_state_qubits is not None and self._breakpoints is None:
+                self.breakpoints = [0, 2 ** num_state_qubits]
+
+            self._reset_registers(num_state_qubits)
+
+    def _reset_registers(self, num_state_qubits: Optional[int]) -> None:
+        if num_state_qubits is not None:
+            qr_state = QuantumRegister(num_state_qubits)
```
Comment (Cryoris) [inline qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:291]: Do we want to give the register names such that it looks nicer when plotting?

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py (hunk 3 above)
Comment (anedumla) [reply qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:175]: changed in the code

File: qiskit/circuit/library/arithmetic/piecewise_chebyshev.py (hunk 4 above)
Comment (anedumla) [reply qiskit/circuit/library/arithmetic/piecewise_chebyshev.py:218]: changed in the code

**Review by Cryoris — APPROVED** [review]

**Pull request merged**